Optionally ignore utf-8 decoding error when converting std::string to python str.

shuminghu opened the pull request:
Summary: When language models use c++ tokenizer, outputs are a c++ strings that are not necessarily valid utf-8 encodings. Default pybind11 casting uses strict utf-8 decoding. We relax the decoding using 'ignore' argument.

Test Plan: https://www.internalfb.com/intern/testinfra/testrun/6473924609918070

Reviewed By: Nayef211

Differential Revision: D43970697

davidberard98 commented on torch/script.h:
```diff
@@ -5,6 +5,7 @@
 #include <torch/csrc/autograd/custom_function.h>
 #include <torch/csrc/autograd/generated/variable_factories.h>
 #include <torch/csrc/autograd/grad_mode.h>
+#include <torch/csrc/jit/python/utf8_decoding_ignore.h>
```
nit: is this necessary?

shuminghu replied:
This is needed to for registering bindings in the follow-up change in [torchtext/csrc/register_torchbindings.cpp](https://github.com/pytorch/text/pull/2126/files#diff-2efc883b84a0f9a8a4abbe2e3da54037ff7d8e6bc4403fd2d7aefde5afa7f044), since it only includes torch/script.h.

davidberard98 replied:
hmm can we just include `torch/csrc/jit/python/utf8_decoding_ignore.h` in register_torchbindings.cpp?

not a big deal but since this is sort of a private api, we probably don't need to include it in torch/script.h (as you can see the current script.h includes a pretty minimal list of .h files)

shuminghu replied:
Unfortunately, it fails with 
```
fbcode/pytorch/text/torchtext/csrc/register_torchbindings.cpp:215:20: error: no member named 'setUTF8DecodingIgnore' in namespace 'torch::jit'
      &torch::jit::setUTF8DecodingIgnore);
 ```
 when `utf8_decoding_ignore.h` is moved from torch/script.h to torchtext/csrc/register_torchbindings.cpp.

davidberard98 replied:
hmm and this is after adding the .h include to both torchtext/csrc/register_pybindings.cpp and torchtext/csrc/register_torchingbindings.cpp

shuminghu replied:
right

davidberard98 reviewed (approved):
LGTM!
- a few nits (inline)
- can you make sure the lint is passing? one of the links shows it's failing but I can't quite find it. (in the pytorch repo, lint failures will get reverted)

shuminghu commented:
>  The lint fix has already been applied. It's from the formatting of adjacent but un-modified code.

Commit:
Optionally ignore utf-8 decoding error when converting std::string to python str. (#97282)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/97282

X-link: https://github.com/pytorch/text/pull/2126

When language models use c++ tokenizer, outputs are a c++ strings that are not necessarily valid utf-8 encodings. Default pybind11 casting uses strict utf-8 decoding. We relax the decoding using 'ignore' argument.

Test Plan: https://www.internalfb.com/intern/testinfra/testrun/4503599786612705

Reviewed By: Nayef211

Differential Revision: D43970697

fbshipit-source-id: d54a7c527d702bec37544442f9ce6e2d441ddc47

shuminghu commented:
@pytorchbot merge
